Rebus.MongoDb2/MongoDbTimeoutStorage.cs

```diff
@@ -57,18 +69,30 @@ public void Add(Timeout.Timeout newTimeout)
         /// </summary>
         public DueTimeoutsResult GetDueTimeouts()
         {
-            var result = collection.Find(Builders<BsonDocument>.Filter.Lte(TimeProperty, RebusTimeMachine.Now()))
-                                   .Sort(Builders<BsonDocument>.Sort.Ascending(TimeProperty));
-
-			return new DueTimeoutsResult(result.Project(r =>
-					new DueMongoTimeout(GetString(r, ReplyToProperty),
-						GetString(r, CorrIdProperty),
-						r[TimeProperty].ToUniversalTime(),
-						GetGuid(r, SagaIdProperty),
-						GetString(r, DataProperty),
-						collection,
-						(ObjectId) r[IdProperty])
-					).ToList());
+            var currentDate = RebusTimeMachine.Now();
+            var dueTimeoutFilter = Builders<BsonDocument>.Filter.Lte(TimeProperty,currentDate);
+            var unlockedTimeoutFilter = Builders<BsonDocument>.Filter.Eq(DueLockProperty, BsonNull.Value) |
+                                Builders<BsonDocument>.Filter.Lt(DueLockProperty, currentDate);
+
+            var timeoutReadyToBeFiredFilter = dueTimeoutFilter & unlockedTimeoutFilter;
+
+            var dueLockShifted = RebusTimeMachine.Now() + _lockTimeoutsOffset;
+
+            var findOneAndUpdateOptions = new FindOneAndUpdateOptions<BsonDocument> { Sort = Builders<BsonDocument>.Sort.Ascending(TimeProperty) };
```

Comment:
Shouldn't you be sorting by DueLock too? (as you added such proparty to existing index previouslly)
If not, why adding it to the index declared early on?

Reply:
Added sort by duelock